In [8]:
print('welcome   python') 

welcome   python


# hello

# Ctrl + Enter, Shift + Enter运行命令
# 选中cell,使用m，即可输入markdown

# 输入dd，删除cell

# 选中cell，输入b,在其后新增cell
# 可以运行javascript脚本

In [6]:
%%js
alert('hello')

<IPython.core.display.Javascript object>

# [ ] 中的数字运行第几次输出
# [*] 正在运行当前cell中的代码中

In [10]:
#注释, 字符串
print('abc'+'efg') #print
print('abc'+"efg") #print

abcefg
abcefg


In [12]:
# 三引号,其中嵌入' ""  多行或特殊字符
"""welcome "to" 
1\
2\
3\




'python' """

'welcome "to" \n123\n\n\n\n\'python\' '

In [15]:
#ipython特殊变量
# _ 下划线，保存上个cell执行结果
_


'welcome "to" \n123\n\n\n\n\'python\' '

In [16]:
#print语句遇到\n会转义
print(_)

welcome "to" 
123



'python' 


In [17]:
#r字符串不转义
print(r"welcome \n magedu")

welcome \n magedu


In [19]:
#换行符：注意换行符后不能接任何字符，应该立即回车

# 运算符

In [20]:
#强和弱类型区别，不同类型是否可以运算
# 强：不能
# 弱：可以隐式转换。

print(1+'a')

TypeError: unsupported operand type(s) for +: 'int' and 'str'

# 自动完成隐式转换不同类型为相同类型

F12

console.log(1+'a')

1a



静态编译： 申明类型，后面不可修改类型


动态编译：不申明，定义即申明类型，可以修改类型

In [23]:
#自然除: 出来浮点型
1/5

0.2

In [25]:
#整除：出来是整数
1//5

0

In [26]:
#取模 %: 取余数
5 % 3

2

In [27]:
# 次方
2**3

8

In [28]:
# 开平方
2**0.5

1.4142135623730951

In [29]:
#开平方
import math
math.sqrt(2)

1.4142135623730951

In [30]:
#~12
bin(12)

'0b1100'

In [32]:
#十进制转换二进制
bin(-1)

'-0b1'

# python返回的是二进制原码转换为十进制数

# 计算机内存中存储的是补码
  * 正数补码是原码本身
  * 负数补码是符号位不变，其余取反+1

~12

12原码 0000 1100

取反  1111 0011

求结果。存储的是补码，正数补码是本身。负数补码是符号位不变，其余取反+1。可以知道这是一个负数

由1111 0011补码求负数，-1除符号位取反。

1000 1101

-13

In [35]:
#!=, ==可以实现不同类型比较, 比较时将两边统一类型。
1 == '1'

False

In [36]:
# 支持连等，但是不要用，引用时出大事。
x = y = z = 10

In [38]:
y

10

# 动态语言，赋值即申明类型，后重新赋值即修改类型。
# 静态语言，先申明类型，后不能修改。

In [40]:
#赋值表达式
b = 6

In [41]:
#引用计数机制：
b = 6 #python一切是对象或实例，6对象引用1
a = b #6引用2
b = 5 #6引用1

# 单分支结构

In [42]:
if 1 < 2:
    print('1 less than 2')

1 less than 2


In [43]:
if 1 > 2:
    print('1 less than 2')

In [44]:
1 > 2

False

In [51]:
bool(-1)

True

# 真值表

In [46]:
a = None

# 多分支

In [54]:
a = 0

if a < 0:
    print('negative')
elif a == 0:
    print('zero')
else:
    print('positive')

zero


# 分支嵌套

In [55]:
score = 80

if score < 0:
    print('wrong')
else:
    if score == 0:
        print('egg')
    elif score <= 100:
        print('right')
    else:
        print('too big')

right


In [59]:
score = 80

if score > 0:
    if score < 100:
        print('right')
    else:
        print('too big')
elif score == 0:
    print('egg')
else:
    print('wrong')

right


In [65]:
score = 80 

if score > 100:
    print('too big')
elif score == 0:
    print('zero')
elif score < 0:
    print('wrong')
else:
    print('right')

right


In [70]:
num = input()

1234


In [71]:
num

'1234'

In [73]:
num = 1234

In [74]:
num

1234

# 给定一个不超过5位的正整数，判断其有几位。例如123

考的是折半思想

if嵌套，可实现折半。if拉平，没有效率。

In [108]:
num = input('Enter a number less than 5 digits: ')
# print(type(num))
num = int(num)

if num / (10**1) < 1:
    print(num, '1')
elif num / (10**2) < 1:
    print(num, '2')  
elif num / (10**3) < 1:
    print(num, '3')
elif num / (10**4) < 1:
    print(num, '4')
else:
    print('over 5 digits')

Enter a number less than 5 digits: 1234
1234 4


In [107]:
num = input('Enter a number less than 5 digits: ')
# print(type(num))
num = int(num)

if num < 0:
    print('wrong')
elif num < 10:
    print(1)
elif num < 100: 
    print(2)
elif num < 1000:
    print(3)
elif num < 10000:
    print(4)
else:
    print('over 5 digits')

Enter a number less than 5 digits: -1
<class 'str'>
wrong


# 折半
计算浪费cpu时间

In [13]:
num = int(input())
data = num/100

if data >= 1:
    if data < 10:
        print(3)
    elif data < 100:
        print(4)
    elif data < 1000:
        print(5)
    else:
        print('over 5')
else:
    data = num/10
    if data >= 1:
        print(2)
    elif data >= 0:
        print(1)
    else:
        print('less than 0')

12345
5


# 老师，折半

In [16]:
val = input('>>> ')
val = int(val)

if val >= 1000: #fold
    if val >= 10000:
        print(5)
    else:
        print(4)
else:
    if val >= 100:
        print(3)
    elif val >= 10:
        print(2)
    else:
        print(1)

>>> 111
3
